mobile-widgets/uploadDiveLogsDE.cpp

```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "uploadDiveLogsDE.h"
+#include <QDir>
+#include <QDebug>
+//#include <QIODevice>
+#include <zip.h>
+#include <errno.h>
+#include "core/display.h"
+#include "core/errorhelper.h"
+#include "core/qthelper.h"
+#include "core/dive.h"
+#include "core/membuffer.h"
+#include "core/divesite.h"
+#include "core/cloudstorage.h"
+#include "core/settings/qPrefCloudStorage.h"
+
+
+uploadDiveLogsDE *uploadDiveLogsDE::instance()
+{
+	static uploadDiveLogsDE *self = new uploadDiveLogsDE;
+
+	return self;
+}
+
+
+uploadDiveLogsDE::uploadDiveLogsDE():
+	reply(NULL),
+	multipart(NULL)
+{
+	timeout.setSingleShot(true);
+}
+
+
+void uploadDiveLogsDE::doUpload(bool selected, const QString &userid, const QString &password)
+{
+	/* generate a temporary filename and create/open that file with zip_open */
+	QString filename(QDir::tempPath() + "/divelogsde-upload.dld");
+
+	// delete file if it exist
+	QFile f(filename);
+	if (f.open(QIODevice::ReadOnly)) {
+		f.close();
+		f.remove();
+	}
+
+	// Make zip file, with all dives, in divelogs.de format
+	if (!prepareDives(selected, filename)) {
+		report_error(tr("Failed to create uploadDiveLogsDE file %s\n").toUtf8(), qPrintable(filename));
+		return;
+	}
+
+	// And upload it
+	uploadDives(filename, userid, password);
+}
+
+
+bool uploadDiveLogsDE::prepareDives(bool selected, const QString &filename)
+{
```

Comment:
do this function is remarkably similar to `DivelogsDeWebServices::prepare_dives_for_divelogs()` except that the arguments are reversed...
I haven't gone through the whole series, just starting the review, but I immediately wonder if it wouldn't make more sense to have a shared class in `core/` to take care of these tasks. Is there something fundamentally different that I haven't noticed, yet?

Reply:
The should not be any relevant differences.